File: path-planning/Bug1&Bug2_algorithm/matlab,C/bug1.c
#include <stdio.h>
// #include <windows.h> // 윈도우
#include <unistd.h> // 리눅스

#define LAND 0
#define OBS 1
#define GOAL 2
#define BUG 3

int MAP[5][5] = {0,};
FILE *fp;

int MAPPlot(int MAP[5][5]);
int* goForward(int *bug);
int motionToGo(int *bug, int *goal);
int* boundaryFollowing1(int *bug, int *goal);
int boundaryFollowing2(int *bug, int *shortest_path);

void dosclear(){
    // system("cls"); // 윈도우
    system("clear"); // 리눅스
}

void dospause(){
    // Sleep(500); // 윈도우
    sleep(1); // 리눅스
}

int main() {
    fp = fopen("bug_history.txt", "w");

    // obs
    int obs_pos[3][2] = {{2,2},{3,2},{1,2}};
    for (int i = 0; i < 3; i++) {
        MAP[obs_pos[i][0]][obs_pos[i][1]] = OBS;
    }

    // goal
    int goal_pos[2] = {2,4};
    MAP[goal_pos[0]][goal_pos[1]] = GOAL;

    // bug (x,y,theta)
    int bug_pos[3] = {1,0,270};
    MAP[bug_pos[0]][bug_pos[1]] = BUG;

    // Bug 1 Algorithm
    int col = 0; // For Check Collision
    int shortest_path[2] = {0,0};
    dosclear();
    MAPPlot(MAP);
    fprintf(fp, "%d %d\n", bug_pos[0], bug_pos[1]);
    dospause();
    
    while (1) {
        if (col == 0) {
            col = motionToGo(bug_pos, goal_pos);
        }

        if (col == 1) {
            int *tmp = boundaryFollowing1(bug_pos, goal_pos); // 한 바퀴 돌면서 가장 가까운 지점 찾기
            shortest_path[0] = tmp[0];
            shortest_path[1] = tmp[1];
            col = boundaryFollowing2(bug_pos, shortest_path); //  가장 가까운 지점으로 이동
        }

        if (col == 2) {
            break;
        }
    }

    printf("도착!\n");
	return 0;
}

int MAPPlot(int MAP[5][5]){
    printf("   0  1  2  3  4\n");
    for (int i = 0; i < 5; i++) {
        printf("%d ", i);
        printf("|");
        for (int j = 0; j < 5; j++) {
            if (MAP[i][j] == 0) printf("  ");
            else if (MAP[i][j] == 1) printf("X ");
            else if (MAP[i][j] == 2) printf("G ");
            else if (MAP[i][j] == 3) printf("B ");
            printf("|");
        }
        printf("\n");
    }
    printf("  ￣￣￣￣￣￣￣￣\n\n");
    return 0;
}

int* boundaryFollowing1(int *bug, int *goal) {
    int original_bug[3];
    original_bug[0] = bug[0];
    original_bug[1] = bug[1];
    original_bug[2] = bug[2];
    
    bug[2] = (bug[2] + 90) % 360; // 우측으로 방향 전환.

    int *shortest_path = (int *)malloc(2 * sizeof(int));
    // int shortest_path[2] = {bug[0], bug[1]}; // 골과 가장 가까운 지점
    shortest_path[0] = bug[0];
    shortest_path[1] = bug[1];

    while (1) {
        dosclear();
        MAPPlot(MAP);
        if ((goal[0] - bug[0]) * (goal[0] - bug[0]) + (goal[1] - bug[1]) * (goal[1] - bug[1]) < 
            (goal[0] - shortest_path[0]) * (goal[0] - shortest_path[0]) + 
            (goal[1] - shortest_path[1]) * (goal[1] - shortest_path[1])) {
            shortest_path[0] = bug[0];
            shortest_path[1] = bug[1];
        }
        printf("shortest_path: %d, %d\n", shortest_path[0], shortest_path[1]);
        
        // 1-step
        int *prevBug = malloc(3 * sizeof(int));
        prevBug[0] = bug[0];
        prevBug[1] = bug[1];
        prevBug[2] = bug[2];

        int *result = goForward(bug); // 전진
        bug[0] = result[0];
        bug[1] = result[1];

        fprintf(fp, "%d %d\n", bug[0], bug[1]);
        dospause();

        MAP[prevBug[0]][prevBug[1]] = LAND; // update the MAP
        MAP[bug[0]][bug[1]] = BUG;

        // temp_bug를 통해 기존 방향으로 이동 가능한지 확인
        int temp_bug[3] = {bug[0], bug[1], original_bug[2]};
        int *temp_result = goForward(temp_bug);
        if (temp_result[2] == LAND) { // 기존 방향으로 이동 가능하면 
            bug[2] = original_bug[2]; // 그 방향을 바라보게 하고
            original_bug[2] = (original_bug[2] - 90 + 360) % 360; // original_dir 갱신(grid world로, 90도별로 방향이 분리되어 있음)
        }

        if (bug[0] == original_bug[0] && bug[1] == original_bug[1]) { // Boundary Following을 시작한 첫 위치에 도착하면,
            bug[2] = original_bug[2]; // 처음 상태와 동일하게 만듦.
            break;
        }
    }
    
    dosclear();
    MAPPlot(MAP);
    printf("shortest_path: %d, %d\n", shortest_path[0], shortest_path[1]);
    dospause();

    return shortest_path;

}

int boundaryFollowing2(int *bug, int *shortest_path) {
    int original_bug[3];
    original_bug[0] = bug[0];
    original_bug[1] = bug[1];
    original_bug[2] = bug[2];
    
    bug[2] = (bug[2] + 90) % 360; // 우측으로 방향 전환.

    int col=0;

    while (1) {
        // 1-step
        int *prevBug = malloc(3 * sizeof(int));
        prevBug[0] = bug[0];
        prevBug[1] = bug[1];
        prevBug[2] = bug[2];

        int *result = goForward(bug);
        bug[0] = result[0];
        bug[1] = result[1];
        col = result[2];

        MAP[prevBug[0]][prevBug[1]] = LAND;
        MAP[bug[0]][bug[1]] = BUG;

        // temp_bug를 통해 기존 방향으로 이동 가능한지 확인
        int temp_bug[3] = {bug[0], bug[1], original_bug[2]};
        int *temp_result = goForward(temp_bug);
        if (temp_result[2] == LAND) { // 기존 방향으로 이동 가능하면 bug 방향 돌려놓기.
            bug[2] = original_bug[2];
            original_bug[2] = (original_bug[2] - 90 + 360) % 360; // original_dir 갱신
        }

        dosclear();
        MAPPlot(MAP);
        fprintf(fp, "%d %d\n", bug[0], bug[1]);
        printf("shortest_path: %d, %d\n", shortest_path[0], shortest_path[1]);
        dospause();

        if (bug[0] == shortest_path[0] && bug[1] == shortest_path[1]) {
            col = 0;
            break;
        }
    }

    return col;
}

int motionToGo(int *bug, int *goal) {
    int prevBug[3];
    prevBug[0] = bug[0];
    prevBug[1] = bug[1];
    prevBug[2] = bug[2];

    // Goal을 향한 방향 맞추기
    int rel_vec[2] = {goal[0] - bug[0], goal[1] - bug[1]};
    if (abs(rel_vec[0]) > abs(rel_vec[1])) {
        if (rel_vec[0] > 0) {
            bug[2] = 0;
        } else {
            bug[2] = 180;
        }
    } else {
        if (rel_vec[1] > 0) {
            bug[2] = 270;
        } else {
            bug[2] = 90;
        }
    }

    // 전진
    int *result = goForward(bug);
    bug[0] = result[0];
    bug[1] = result[1];
    int col = result[2];

    if (col == OBS) { // Return to prev state.
        bug[0] = prevBug[0];
        bug[1] = prevBug[1];
    } else { // Update the MAP.
        MAP[prevBug[0]][prevBug[1]] = LAND;
        MAP[bug[0]][bug[1]] = BUG;

        dosclear();
        MAPPlot(MAP);
        fprintf(fp, "%d %d\n", bug[0], bug[1]);
        dospause();
    }

    return col;
}

int* goForward(int *bug) {
    int *result = malloc(3 * sizeof(int));

    if (bug[2] == 0) {
        bug[0]++;
    } else if (bug[2] == 90) {
        bug[1]--;
    } else if (bug[2] == 180) {
        bug[0]--;
    } else if (bug[2] == 270) {
        bug[1]++;
    }

    result[0] = bug[0];
    result[1] = bug[1];

    if (MAP[bug[0]][bug[1]] == LAND) {
        result[2] = 0; // LAND
    } else if (MAP[bug[0]][bug[1]] == OBS) {
        result[2] = 1; // OBS
    } else if (MAP[bug[0]][bug[1]] == GOAL) {
        result[2] = 2; // GOAL
    }

    return result;
}

File: path-planning/Bug1&Bug2_algorithm/matlab,C/bug1.m
clc; close all; clear all;

% Environment Settings
% Map(0: 땅, 1: 벽돌, 2: 도착점, 3: 버그)
global LAND OBS GOAL BUG;
LAND = 0;
OBS = 1;
GOAL = 2;
BUG = 3;

mapSize = linspace(1, 5, 5);
map=zeros(length(mapSize));

% obs
obs=[3 3;
    4 3
    2 3];
map(obs(:, 1), obs(:, 2)) = OBS;

% goal
goal = [3, 5];
map(goal(1), goal(2)) = GOAL;

% bug (x, y, theta)
bug = [2 1 270];
map(bug(1), bug(2)) = BUG;


% Bug 1 Algorithm.
col = 0;                % For Check Collision
shortest_path = [0 0];

MapPlot(map);
pause(0.5);

while true
    if col == 0
        disp("Motion To Go 시작!")
        [map, bug, col] = MotionToGo(map, bug, goal);
    end
    
    if col == 1
        disp("Boundary Following 시작!")
        [map, bug, shortest_path] = BoundaryFollowing1(map, bug, goal);
        [map, bug, col] = BoundaryFollowing2(map, bug, shortest_path);
    end
    
    if col == 2
        break
    end

    
end
disp("도착!")


% Functions
function [map, bug, shortest_path] = BoundaryFollowing1(map, bug, goal)
    global LAND BUG
    original_bug = bug;     % 처음 위치 저장.
    bug(3) = mod(bug(3) + 90, 360);             % 우측으로 방향 전환.
    
    shortest_path = bug(1:2);

    % 바운더리 팔로잉 시작.
    while true
        if norm(goal(1:2) - bug(1:2)) < norm(goal(1:2) - shortest_path)
            shortest_path = bug(1:2);
        end

        % 1-step      
        prevBug = bug;
        [bug, col] = GoForward(bug, map);           % 전진.
        map(prevBug(1), prevBug(2)) = LAND;         % Update the Map.
        map(bug(1), bug(2)) = BUG;
    
        % temp_bug를 통해 기존 방향으로 이동 가능한지 확인
        temp_bug = bug;                             
        temp_bug(3) = original_bug(3);
        [temp_bug, col] = GoForward(temp_bug, map);
        if col == 0                                 % 기존 방향으로 이동 가능하면 bug 방향 돌려놓기.
            bug(3) = original_bug(3);
    
            original_bug(3) = original_bug(3) - 90;   % original_dir 갱신
            if original_bug(3) < 0 
                original_bug(3) = 270;
            end
        end

        MapPlot(map);
        pause(0.5);

        if bug(1:2) == original_bug(1:2)        % Boundary Following을 시작한 첫 위치에 도착하면,
            bug(3) = original_bug(3);           % 처음 상태와 동일하게 만듦.
            break;
        end      
    end
end

function [map, bug, col] = BoundaryFollowing2(map, bug, shortest_path)
    global LAND BUG
    original_bug = bug;
    bug(3) = mod(bug(3) + 90, 360);             % 우측으로 방향 전환.
    
    % 바운더리 팔로잉 시작.
    while true
        % 1-step      
        prevBug = bug;
        [bug, col] = GoForward(bug, map);           % 전진.
        map(prevBug(1), prevBug(2)) = LAND;         % Update the Map.
        map(bug(1), bug(2)) = BUG;
    
        % temp_bug를 통해 기존 방향으로 이동 가능한지 확인
        temp_bug = bug;                             
        temp_bug(3) = original_bug(3);
        [temp_bug, col] = GoForward(temp_bug, map);
        if col == 0                                 % 기존 방향으로 이동 가능하면 bug 방향 돌려놓기.
            bug(3) = original_bug(3);
    
            original_bug(3) = original_bug(3) - 90;   % original_dir 갱신
            if original_bug(3) < 0 
                original_bug(3) = 270;
            end
        end

        MapPlot(map);
        pause(0.5);

        if bug(1:2) == shortest_path
            col = 0;
            break;
        end
    end
end

function [bug, col] = GoForward(bug, map)
    global LAND OBS GOAL
    if bug(3) == 0
        bug(1) = bug(1) + 1;
    elseif bug(3) == 90
        bug(2) = bug(2) - 1;
    elseif bug(3) == 180
        bug(1) = bug(1) - 1;
    elseif bug(3) == 270
        bug(2) = bug(2) + 1;
    end

    if map(bug(1), bug(2)) == LAND
        col = 0;
    elseif map(bug(1), bug(2)) == OBS
        col = 1;
    elseif map(bug(1), bug(2)) == GOAL
        col = 2;
    end
end

function [map, bug, col] = MotionToGo(map, bug, goal)
    global LAND OBS GOAL BUG;
    prevBug = bug;                      % 이전 버그 위치
    
    % Goal을 향한 방향 맞추기
    rel_vec = goal(1:2) - bug(1:2);
    if abs(rel_vec(1)) > abs(rel_vec(2))
        if rel_vec(1) > 0
            bug(3) = 0;
        else
            bug(3) = 180;
        end
    else
        if rel_vec(2) > 0
            bug(3) = 270;
        else
            bug(3) = 90;
        end
    end
   

    [bug, col] = GoForward(bug, map);   % 전진

    if col == OBS
        bug = prevBug;                  % Return to prev state.

    elseif col == LAND
        map(prevBug(1), prevBug(2)) = LAND;        % Update the Map.
        map(bug(1), bug(2)) = BUG;

    elseif col == GOAL
        map(prevBug(1), prevBug(2)) = LAND;        % Update the Map.
        map(bug(1), bug(2)) = BUG;
    end

    MapPlot(map);
    pause(0.5);
end

function MapPlot(map)
    map_image = imresize(map, 40, "nearest");
    imshow(map_image, Colormap=jet);
    caxis([0 3])
    xlabel("Y")
    ylabel("X")
end
